ping-payload-extensions/ping-custom-payload-extensions.md

```diff
@@ -0,0 +1,104 @@
+```python
+
+# Max ASCII hex encoded strings length
+MAX_ERROR_BYTE_LENGTH = 300
+
+error_payload = SSZ.serialize(Container(error_code: u16, message: ByteList[MAX_ERROR_BYTE_LENGTH]))
+
+ErrorPayload = Container(
+  type: 65535,
+  payload: error_payload
+)
+```
+
+### Error Code's
+
+#### 0: Extension not supported
+This code should be returned if the extension isn't supported. This error should only be returned if 
+- The extension isn't supported
+- The extension isn't a required extension for specified Portal Network.
+
+#### 1: Requested data not found
+This error code should be used when a client is unable to provide the necessary data for the response.
+
+#### 2: Failed to decode payload
+Wasn't able to decode the payload
+
+#### 3: System error
+A critical error happened and the ping can't be processed
+
+## Standard extensions
+
+Every Portal Network Client is required to support 3 extensions
+- [Type 0: Capabilities Payload](extensions/type-0.md): useful for finding what ping extensions a client supports
+- Sub-network standard extension. For history this is currently [Type 2: History Radius Payload](extensions/type-2.md). For State and Beacon the base extension is [Type 1: Basic Radius Payload](extensions/type-1.md). These are extensions what are crucial. This standard extensions are specified the respective sub-networks specification  
+- Type 3: [Get Client info Payload](extensions/type-3.md)
+
+## What is the [Type 0: Capabilities Payload](extensions/type-0.md) for
+It is for Portal implementations which want to see what extensions a peer supports. Not all extensions need to be implemented by all parties. So in order for a peer to find if an extension is implemented a [Type 0: Capabilities Payload](extensions/type-0.md) should be exchanged.
+
+Non-required extension offer a way for Portal implementations to offer extended functionality to its users without requiring every Portal implementing party to agree to a new feature. This allows for a diverse set of use cases to be fulfilled without requiring every implementer implement every extension, or requiring the need to bloat the minimal [Portal-Wire-Protocol](../portal-wire-protocol.md) with new `Message Types`.
+
+## Does implementing non-standard-extensions require a hardfork?
+No only changing a sub-networks standard extension requires a hard fork.
+
+
+## How do sub-network standard extension's work
+sub-network standard extension's are fundamental extensions that are required for a Portal sub-network to function. They must be supported by the sub-networks implementations.
+
+### What is the process for changing a sub-network standard extension
+
+The process of changing a Portal sub-networks standard extension is through a hard fork set by a timestamp. When the timestamp is reached usage of the old standard extension is replaced by the new standard extension. Details on what standard extensions for what networks are being replace are specified in the fork's document.
+
```

Comment:
For me the usefulness of these ping/pong payload extensions is to avoid having to do something like this in the first place.

If we don't avoid this hard-forking, then I don't really understand the point of adding this complexity.

But imo it is possible to do it without hard-forking, as is also explained in the chapter below this?

Reply:
I would suggest that we remove this section and leave the process undefined.  I think we are getting into the realm of over-specifying this piece of the protocol.

Reply:
A client would never need to call `capabilities` to know what the current standard extension is.

The usefulness of ping extensions is not requiring all parties to agree to make a change to the protocol or add a feature.

Of course supporting the base protocol is required i.e. `standard extensions`, but this allows Portal clients to make optimizations such as the "welcome basket" proposal  https://github.com/ethereum/portal-network-specs/issues/350 Carver proposed, without requiring all implementations to support it. Which is important as we don't want new features to gate keep new clients from joining the network, but also not all clients would likely support "welcome baskets" it is a p2p optimization which isn't required by the core protocol to work. It is a bootstrapping optimization.

Ping extensions don't change the base protocol much, new clients can just return empty for `capabilities` and only support X networks `standard extension`. `Ping Extensions` is opt in and hence it prevents creating a situation of the protocol being "too tall to ride( or participate/implement a client)."

If a client wants a feature what is not popular or doesn't make sense for all clients to implement it they can, which I think is important long term. Especially as not only L1 will be using Portal. The base Portal P2P protocol should be simple, and Ping extensions allow us to keep it simple. Well still allowing for clients to experiment or add new features, which maybe one day become standard extensions if they become popular, but there is no requirement for that, which is one of the important parts of this proposal.

Reply:
My comment is not about the usefulness of the extensions. It is about this specific section which talks about the requirement of a hard fork.

I think this should be designed in such way that a hard fork is not required at all, else I don't see the point of it, and with the capabilities extension this should be feasible.

Reply:
Are we assuming
1. ping's are only done by long-lived connections i.e. the other node is in your routing table? All short lived connections are done through `find nodes` or `find content` RFC/RFN etc

or 

2. ping's are sent by both long-lived and short-lived connections. 

?

If 1. is true I think a forkless mechanism makes sense as the cost of calling `capabilities` is only high if the connection is `shortlived`. If ping isn't used for `short-lived connections`, then the additional latency incurred isn't a big deal.

if 2. is true I think the `fork upgrade method for standard extensions make sense` as it wouldn't make sense to upgrade via `capabilities` unless

- So for "close nodes" upgrading to the newest protocol or using extensions could make sense as those will be long lived connections in the routing table
- if you are trying to request meta data, calling `capabilities` would make sense
- participate in features not supported by the full node set

but if you were a `short-lived ping connection`, upgrading wouldn't make sense, only using the base standard extension required by the protocol would make sense

I need to put more thought into this as these are only my additional thoughts. I didn't conceive 1. could be an option till just now so I still need to spend time thinking which assumption is actually true 1. or 2.

A forkless method makes a lot of sense for `devp2p` as they only have long-lived connections. Since Portal is by default a `short-lived` protocol, having a `forked upgrade method for the base standard extensions required by the protocol` makes sense if 2. is true, as you wouldn't want to incur the latency for short-lived connections.

Reply:
I guess there is more thinking to be done, some of this we would only need to worry about when we propose the first breaking change so I don't think we need to pick a path immediately 

The 2 paths I see are

1. We always require `capabilities` is called first to find the latest shared version of the respective subnetworks `standard extension`. This path would be `forkless`, but it would always occur 1 additional round trip delay initially to make a `standard extension` ping, as there could be multiple support `standard extension` in this system

2. We do a `fork` system, where all client implementations would always support a base `standard extension` for the subnetwork to work/function. This system would make it optional if an implementation wanted to call `capabilities` or not and hence would not have the additional RT time unless opted into.

Both have trade offs, to ask bluntly which path do you want to do?